### PR TITLE
Add a transformer neural operator model and an accompanying example for training it on Darcy flow

### PR DESCRIPTION
**Pull request by zijieli-Jlee**
Main changes:

* Add `transformer_no.py` under `models`, which is a Transformer encoder-decoder architecture for learning operator.  Small changes to the `__init__.py` and `model_dispatcher.py` accordingly.
* Add Gaussian Fourier feature and Siren to the `embeddings.py` which is useful for building the neural field (query point embedding) in the Transformer decoder.
* Add a training example `train_darcy_transformer.py` under `scripts`, the corresponding config yaml file is `darcy_transformer_config.yaml`.
* Fixed a bug in previous attention layer test [#290](https://github.com/neuraloperator/neuraloperator/issues/290)

### CODE REVIEW

**Review by dhpitt — COMMENTED** [review]
Thanks @zijieli-Jlee! This looks awesome. A few minor things:

1.  Thanks for fixing #290 ! Would it be OK to split up this PR into one that fixes the issue (which we can merge immediately) and one that implements the TransformerNO (which may take longer to go over)?
2. It would be great to add unit tests for the `TransformerNO` model, along the lines of `neuralop/models/tests/test_fno.py` (just verifying that the forward pass produces outputs that we would expect)
3. It might help keep the code flexible and easier to maintain if the `EncoderBlocks` was its own module in `neuralop.layers`, along the lines of the `FNOBlocks`.

File: scripts/train_darcy_transformer.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    n_tests=config.data.n_tests,
+    test_batch_sizes=config.data.test_batch_sizes,
+    encode_input=config.data.encode_input,
+    encode_output=config.data.encode_output,
+)
+
+model = get_model(config)
+
+
+class ModelWrapper(torch.nn.Module):
```
Comment (dhpitt) [inline scripts/train_darcy_transformer.py:85]: What's the function of the ModelWrapper class? It seems like it adds functionality to the Transformer NO's forward pass. Some documentation would be helpful.

File: neuralop/models/transformer_no.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+            raise ValueError(f'pos_emb must be one of ["rotary", "none"], got {self.pos_emb}')
+
+        if self.pos_emb == 'rotary':
+            self.enc_pos_emb_module = RotaryEmbedding(self.encoder_head_n_channels // self.n_dim)
+            self.dec_pos_emb_module = RotaryEmbedding(self.decoder_head_n_channels // self.n_dim)
+        else:
+            self.enc_pos_emb_module = None
+            self.dec_pos_emb_module = None
+
+        self.encoder = nn.ModuleList(
```
Comment (dhpitt) [inline neuralop/models/transformer_no.py:117]: nit: `FNOBlocks` takes `n_layers` as a parameter instead of exposing a moduleList within the `FNO` itself. It might be nice to have a similar convention here

**Review by dhpitt — COMMENTED** [review]
Thank you for the updates @zijieli-Jlee, the PR is in really good shape overall. I left a few small comments

**Comment by zijieli-Jlee** [comment]
Hi @dhpitt , thanks for the comments!

1. I saw this was fixed at #295 , let me know if there is still any remaining issue
2. Definitely. I will add the test for the `TransformerNO` if its current structure looks fine (see the next point)
3. Thanks for the suggestion! I added `transformer_block` following `fno_block` to `layers`, the `transformer_block` contains implementation for `TransformerEncoderBlock` and `TransformerDecoderBlock`. I also added test for these modules to `test_transformer_block`

File: neuralop/models/transformer_no.py (hunk 2 above)
Comment (zijieli-Jlee) [reply neuralop/models/transformer_no.py:117]: Got it;)

**Comment by dhpitt** [comment]
@zijieli-Jlee thanks for the edits, your PR is in good shape. if you fix the conflicts we can re-run the tests and see if it's ready to go.

**Comment by dhpitt** [comment]
Thanks for the fixes @zijieli-Jlee ! I fixed a one-line conflict with a model import since the last update to main. 

This PR is looking great overall, but I think your layers and util functions could use clearer documentation in the docstrings - see `neuralop/layers/integral_transform.py` for an example of informative docs.

**Comment by zijieli-Jlee** [comment]
I added docstring to the transformer_block and some newly added positional encoding class. In addition I just spotted and fixed a bug in the normalization inside attention layer

File: neuralop/layers/transformer_block.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+                ):
+        """
+           Project the input function u from the source grid to the query grid using the Transformer Decoder Block.
+
+          Parameters:
+                u: torch.Tensor, input tensor of shape [batch_size, num_src_grid_points, channels]
+                pos_src: torch.Tensor, grid point coordinates of shape [batch_size, num_src_grid_points, channels]
+                pos_emb_module: nn.Module, positional embedding module, by default None
+                pos_qry: torch.Tensor, grid point coordinates of shape [batch_size, num_sry_grid_points, channels],
+                         by default None and is set to pos_src
```
Comment (dhpitt) [inline neuralop/layers/transformer_block.py:252]: It would be useful to include the specific purpose of `pos_qry` as compared to pos_src - inputs to the query basis function

File: neuralop/layers/embeddings.py [hunk 4]
```diff
@@ -73,4 +73,173 @@ def apply_2d_rotary_pos_emb(t, freqs_x, freqs_y):
...
 
         return torch.cat((apply_rotary_pos_emb(t_x, freqs_x),
-                          apply_rotary_pos_emb(t_y, freqs_y)), dim=-1)
\ No newline at end of file
+                          apply_rotary_pos_emb(t_y, freqs_y)), dim=-1)
+
+
+# Gaussian random Fourier features
+# code modified from: https://github.com/ndahlquist/pytorch-fourier-feature-networks
+class GaussianFourierFeatureTransform(nn.Module):
```
Comment (dhpitt) [inline neuralop/layers/embeddings.py:81]: nit: these comments probably belong in the docstring

File: neuralop/layers/embeddings.py [hunk 5]
```diff
@@ -73,4 +73,173 @@ def apply_2d_rotary_pos_emb(t, freqs_x, freqs_y):
...
+        # From [(B*N), C] to [B, N, C]
+        x = x.view(batches, num_of_points, -1)
+
+        x = 2 * torch.pi * x
+
+        return torch.cat([torch.sin(x), torch.cos(x)], dim=-1)
+
+
+# SirenNet
+# code modified from: https://github.com/lucidrains/siren-pytorch/blob/master/siren_pytorch/siren_pytorch.py
```
Comment (dhpitt) [inline neuralop/layers/embeddings.py:127]: same as above - cleaner to leave these comments in the docstring

**Review by dhpitt — COMMENTED** [review]
This PR is looking great. My only comment is a small nitpick about putting the model source within docstrings. Otherwise I think this is ready to go.

**Comment by dhpitt** [comment]
Looks great! Asking @JeanKossaifi  for final approval.

File: scripts/train_darcy_transformer.py (hunk 1 above)
Comment (mliuschi) [reply scripts/train_darcy_transformer.py:85]: Is it possible to incorporate the things in `ModelWrapper` into the dataset instead? I think this has been the convention we have been taking in the library. @JeanKossaifi Feel free to correct me if I'm wrong.

File: neuralop/layers/transformer_block.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+
+        """
+        u = self.lifting(u)
+        for l in range(self.n_layers):
+            u_attention_skip = u
+            u = self.attention_layers[l](u_src=normalize(u, self.attention_norms[l]),
+                                     pos_src=pos,
+                                     positional_embedding_module=pos_emb_module,
+                                     **kwargs)
+            u = u + u_attention_skip
```
Comment (mliuschi) [inline neuralop/layers/transformer_block.py:134]: Have you noticed any empirical improvements from having this skip connection be a pointwise linear (not MLP) as opposed to identity connection? If so, it could be interesting to supplement with this functionality.

File: scripts/train_darcy_transformer.py (hunk 1 above)
Comment (mliuschi) [reply scripts/train_darcy_transformer.py:85]: Is it possible to incorporate the things in `ModelWrapper` into the dataset instead? I think this has been the convention we have been taking in the library. @JeanKossaifi Feel free to correct me if I'm wrong.

**Review by mliuschi — COMMENTED** [review]
Left a couple of comments only because it mostly looks good to me!

File: neuralop/layers/embeddings.py [hunk 7]
```diff
@@ -73,4 +74,170 @@ def apply_2d_rotary_pos_emb(t, freqs_x, freqs_y):
...
+        """
+            code modified from:
+                 https://github.com/lucidrains/siren-pytorch/blob/master/siren_pytorch/siren_pytorch.py
+            SIREN paper: https://arxiv.org/abs/2006.09661
+            The Siren layer is a linear layer followed by a sine activation function.
+
+            Parameters:
+                dim_in: int, Number of input channels.
+                dim_out: int, Number of output channels.
+                w0: float, scaling factor (denominator) used to initialize the weights, by default 6.
```
Comment (juliusberner) [inline neuralop/layers/embeddings.py:153]: `w0` seems to default to `1` instead of `6`?

Comment (juliusberner) [reply neuralop/layers/embeddings.py:153]: It is also used to initialize the `Sine` activation scaling, is this intentional?

File: neuralop/layers/embeddings.py [hunk 8]
```diff
@@ -73,4 +74,170 @@ def apply_2d_rotary_pos_emb(t, freqs_x, freqs_y):
...
+
+        """
+
+        super().__init__()
+        self.dim_in = dim_in
+        self.is_first = is_first
+
+        weight = torch.zeros(dim_out, dim_in)
+        bias = torch.zeros(dim_out) if use_bias else None
+        self.init_(weight, bias, c=c, w0=w0)
```
Comment (juliusberner) [inline neuralop/layers/embeddings.py:167]: Is there a reason that we do not use `torch.nn.Linear` and `torch.nn.init.uniform_` to initialize?

File: neuralop/models/base_model.py
Comment (juliusberner) [inline neuralop/models/base_model.py:1]: These edits could be removed from the PR

File: neuralop/layers/embeddings.py [hunk 9]
```diff
@@ -73,4 +74,170 @@ def apply_2d_rotary_pos_emb(t, freqs_x, freqs_y):
...
+        if bias is not None:
+            bias.uniform_(-w_std, w_std)
+
+    def forward(self, x):
+        out = torch.nn.functional.linear(x, self.weight, self.bias)
+        out = self.activation(out)
+        return out
+
+
+class SirenNet(nn.Module):
```
Comment (juliusberner) [inline neuralop/layers/embeddings.py:188]: @dhpitt Should we move this to `layers/mlp.py`?

File: neuralop/layers/embeddings.py [hunk 10]
```diff
@@ -73,4 +74,170 @@ def apply_2d_rotary_pos_emb(t, freqs_x, freqs_y):
         t_x, t_y = t[..., :d//2], t[..., d//2:]
 
         return torch.cat((apply_rotary_pos_emb(t_x, freqs_x),
-                          apply_rotary_pos_emb(t_y, freqs_y)), dim=-1)
\ No newline at end of file
+                          apply_rotary_pos_emb(t_y, freqs_y)), dim=-1)
+
+
+class GaussianFourierFeatureTransform(nn.Module):
```
Comment (juliusberner) [inline neuralop/layers/embeddings.py:80]: @dhpitt We now have two Fourier feature embeddings, i.e., `PositionalEmbedding` above and this one --- I think we should name them more consistently, e.g., `GaussianFourierEmbedding` and `FourierEmbedding`?

File: neuralop/layers/embeddings.py (hunk 9 above)
Comment (dhpitt) [reply neuralop/layers/embeddings.py:188]: That makes sense to me

File: neuralop/layers/embeddings.py (hunk 10 above)
Comment (dhpitt) [reply neuralop/layers/embeddings.py:80]: Good idea @zijieli-Jlee

**Review by dhpitt — APPROVED** [review]
This all looks great. @JeanKossaifi

File: neuralop/layers/embeddings.py [hunk 11]
```diff
@@ -73,4 +74,50 @@ def apply_2d_rotary_pos_emb(t, freqs_x, freqs_y):
...
+    def forward(self, x):
+        if len(x.shape) == 2:
+            x = x.unsqueeze(0)
+        batches, num_of_points, channels = x.shape
+
+        # Make shape compatible for matmul with _B.
+        # From [B, N, C] to [(B*N), C].
+        x = x.view(-1, channels)
+
+        x = x @ self._B.to(x.device)
```
Comment (juliusberner) [inline neuralop/layers/embeddings.py:116]: We can remove the `.to(...)` since `self._B` is registered as parameter and automatically transferred with the model.

File: scripts/train_darcy_transformer.py [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+    test_resolutions=config.data.test_resolutions,
+    n_tests=config.data.n_tests,
+    test_batch_sizes=config.data.test_batch_sizes,
+    encode_input=config.data.encode_input,
+    encode_output=config.data.encode_output,
+)
+
+model = get_model(config)
+
+### The following wrapper is to maintain a consistent input format with other FNO-based model
```
Comment (juliusberner) [inline scripts/train_darcy_transformer.py:84]: Is there a reason why we do not directly use the FNO input format for the `TransformerNO`?

**Review by juliusberner — APPROVED** [review]
LGTM, I just added two minor comments.

File: scripts/train_darcy_transformer.py (hunk 12 above)
Comment (JeanKossaifi) [reply scripts/train_darcy_transformer.py:84]: @dhpitt seems like we'd want to use the refactored 2DGrid here instead.

**Comment by JeanKossaifi** [comment]
@zijieli-Jlee this looks great, I agree with @juliusberner it would just be great to make the input the same as FNO and obviate the need for the wrapper :) 
We can also use the positional embedding already existing in the lib!

**Comment by dhpitt** [comment]
Just resolved merge conflicts with changes in `neuralop.layers.embeddings`

File: neuralop/layers/transformer_block.py [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+                **kwargs
+                ):
+        """
+           Project the input function u from the source grid to the query grid using the Transformer Decoder Block.
+
+          Parameters:
+                u: torch.Tensor, input tensor of shape [batch_size, num_src_grid_points, channels]
+                pos_src: torch.Tensor, grid point coordinates of shape [batch_size, num_src_grid_points, channels]
+                pos_emb_module: nn.Module, positional embedding module, by default None
+                pos_qry: torch.Tensor, grid point coordinates of shape [batch_size, num_sry_grid_points, channels],
```
Comment (juliusberner) [inline neuralop/layers/transformer_block.py:291]: nit: `sry` -> `qry`

**Comment by dhpitt** [comment]
Converting this PR to draft until next release.

**Comment by dhpitt** [comment]
Stale; closing
